Async stack traces (structure version)

clementbera opened the pull request:
In this commit I introduce the setting --async_stack_traces_structure and the method asyncError: in SOMns. If enabled, asyncError: will display the stack trace included the stack trace on other actors, even if the stack has changed after the asynchronous call.

This implementation extends the calling convention by adding an extra parameter to each send, building a tree structure representing the call graph. Each send is effectively performing an extra allocation.

smarr committed:
Adopt new interface for before/after root node

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Added support for Truffle’s ExpressionTag

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use Truffle’s Tag root class for tags

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Change in formatting

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use parameter to step out, instead of doing it manually

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Adopt InstrumentableNode/GenerateWrapper and hasTag

- all InstrumentableNodes need execute method with virtual frame for correct wrapping
- need no-arg standard constructor
- replace VM.insertInstrumentationWrapper(.) with notifyInserted(.)
- adopt IsValue primitive
- use hasTag method from InstrumentableNode
- eliminate all isTaggedWith, this is not compatible with the other changes, need to use hasTag exclusively
- fix node initialization and set missing source sections

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Updated to Truffle 1.0.0-rc5, incl. mx, and BD

- updated Maven URL in BD, to prevent 403 error on Travis
- updated mx with Google Maven mirror to prevent Travis issues

- added new DSL processors to Eclipse project
  - updated Maven URL in BD, to prevent 403 error on Travis
  - updated mx with Google Maven mirror to prevent Travis issues

- added new DSL processors to extension test
- updated build dependencies in general
  - make sure Truffle’s tools are built

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove insertInstrumentationWrapper

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Improve Primitive.toString(.) and add others for debugging.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Adopt Instrumenter.attachExecutionEventFactory(.)

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fixed reference to wrong class as tag in ActivityJoin

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update interop basics and turn SAbstractObject into TruffleObject

With the latest truffle changes, all objects going to the instrumentation need to be interop-compatible.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Ignore nodes without source sections

With the latest truffle changes, source sections do not seem to be mandatory anymore, but we need only those nodes.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Optimize tracing

- we know it is a TracingActor at this point, no need for a virtual call to `getId()`
- fix seemingly incorrect sync on fullBuffers
- prefer isEmpty over size()
- fix check for whether there is something in the buffer
- position() should be the correct thing, untested though
- use unsafe to avoid separate byte writes
- avoid synchronizing on a synchronized data structure
- add additional methods to Unsafe dummy.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Prototype for TraceActorCreation node

Don’t write currentActor in every actorContext recording

This seems just in the wrong place, and just writes itself, even if not necessary.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add second node for TraceActorContext

This one is needed for all buffer changes

Remove recordActorCreation, was replaced by node

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove message tracing intro truffle method to enable specialization

Reduce code duplication in received message handling

Move more of the old tracing also intro the received call targets

Factor our redundant code

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add node suffix to node classes for consistency

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Added separate nodes to record ids

This avoids many recurring specializations in the tracing nodes we need.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Refactor tracing nodes to use id nodes

[Fix] id tracing, remove unneeded code

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Turn message tracing into node

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add profile for whenResolved classes

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add separate skip.truffle for ant

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Make SArray constructor protected

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add support for statistics on tracing

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix NPE in uncaught exception handler, and add info on thread

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Move actor execution loop over mailbox into truffle

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Replace compilation warning with boundaries

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove cache for messages in exec loop

- this does not seem to be beneficial (not extensively benchmarked though)
- problem is that we get too many different call targets
- so, looking through list of cached items becomes expensive

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Don’t trace actor context twice on buffer swap

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
AsyncHarness: convert Warmup into simply ignoring results for aggregate statistics

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove extra indirection of byte buffer

ByteBuffer didn’t do much useful things, and cost another indirection for tracing. Let’s see whether this makes any difference.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Optimize scheduling promises

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove All.ns

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Harness: convert Warmup into simply ignoring results for aggregate statistics

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update expected results

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Adapt codespeed.conf for warmup semantics changes

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Make StreamTests independent from current directory

- fix style issues, too

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix style

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Avoid hardcoding test suites

Use the file API to actually read the *Tests.ns files from the directory.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

clementbera committed:
Attempted to fix the issue with Dominik by creating an ExceptionSignalingTruffleNode

- added JUnits tests for Files

Co-authored-by: Dominik Aumayr <daumayr162@gmail.com>

smarr committed:
Add support for extension modules

Extension modules are jar files, and can be loaded like normal modules.
This means, we get back a class, which can be instantiated.
The methods on the class are defined by the @Primitive annotations on node classes.
These primitives have to be made available via the list of factories to the primitive loader.

- added basic tests
- added test extension
- added extension primitive loader
- some refactoring of mixin builder and parser
- added Eclipse and Ant setup
- added documentation

Added extension test Eclipse and Ant setup

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Improve error message in Class>>#new

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update test data

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Whitespace

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Refactor exception signaling further

- pre-allocate commonly used exception-related symbols, and use instead
  of strings
- rename execute to signal, and use directly
- consistently use factory methods for exception nodes
  - reduce extensive verbosity, they are static methods,
    context is provided by the class name
  - change name of parameters: they are selectors,
    could be other things beside classes
  - restructure to enable lazy initialization on first activation
    - this is necessary to resolve modules which are loaded after
      parsing a method
    - also allocate the message send nodes lazily to avoid impact on
      startup performance

- improve source section used for attribution to be the initializer
- use node initializer methods consistently
- remove ValuePrim, simply pass node instead
- avoid `thrower` as name, instead use the type of the exception
  that’s going to be throw. Reads nicer: `argumentError.signal(…)`

- remove unused code

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update test data

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Added missing callerNeedsToBeOptimized to SClass.isKindOf

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Test and document JAR classpath dependencies

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Determine environment dynamically

Signed-off-by: Stefan Marr <git@stefan-marr.de>

daumayr committed:
Fixes for tracing external data, naming, and Kompos

fniephaus committed:
Use CompilerDirectives::castExact

instead of ValueProfiles. The assumption is that performance should remain the same (benchmarks needed for verification). At the same time, this change would simplify code.

Closes #255

fniephaus committed:
Use direct casts instead of `exactCast`

According to @woess, `CompilerDirectives::castExact` is only needed if the target class is not final. Since primitive types as well as PartiallyEmptyArrays are final, the `storage` field can be casted directly. This is not the case for Object[], so use a single ValueProfile until SOMns' Truffle version supports `castExact`.

smarr committed:
Update codespeed.conf for ReBench 1.0

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update to Checkstyle 8.11

This accounts also for their move to GitHub.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update Kompos dependencies

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Don’t use debug mode for rebench

Should not be necessary anymore

Signed-off-by: Stefan Marr <git@stefan-marr.de>

daumayr committed:
Cleanup in Files

make stream creation simpler

daumayr committed:
Remove -atcfg flag and add -sid flag for short ids

The old actor config flags are not used any longer.

daumayr committed:
Add notion of external messages for replay

External messages are going to be use to represent information coming in from outside of the system.

The capture the external non-determinism.
And adds data wrappers for some of the payload.

This is just an initial change, they are not yet used.

daumayr committed:
Rename MEDEOR_TRACING to KOMPOS_TRACING

Use the name of the protocol that is supported by this, not the subsystem that implements it.

daumayr committed:
Split Tracing into Kompos and replay support

daumayr committed:
Update tracing and replay

daumayr committed:
Trace system time and add datasource interface

daumayr committed:
Make buffers configureable

- add options for constants
- make buffer recycling optional

daumayr committed:
Fix buffer handling when assertions are disabled

smarr committed:
Added a microbenchmark for String>>#charAt:

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Added benchmark for basic file reading/writing

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Optimize file reading/writing for compilation

This is only addressing the basic compilation issues, it does not do any performance tuning.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update test data

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[Travis] caskroom is deprecated

Reinstall of Java failed with:

==> Downloading http://download.oracle.com/otn-pub/java/jdk/10.0.2+13/19aef61b38124481863b1413dce1855f/jdk-10.0.2_osx-x64_bin.dmg
Error: undefined method `undent' for #<String:0x000000010233a010>

Hope this fixes it

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Move ActorExecutionTrace to replay package

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix expected test results

Signed-off-by: Stefan Marr <git@stefan-marr.de>

daumayr committed:
Implement handling of TwoDArrayWrapper

daumayr committed:
Fix packages and imports

daumayr committed:
Update test data

smarr committed:
Update to mx with support for compilation from tar ball

- change ant check for submodule to work also for tar ball

Signed-off-by: Stefan Marr <git@stefan-marr.de>

daumayr committed:
Redesign Timer Prim for Tracing

daumayr committed:
Enable Validation Tests

daumayr committed:
Update Test Data

smarr committed:
Replay tests are required to pass, don’t require >=Java9

The setup for Java9 or later does not pass currently, because mx and our Truffle version are at this point not compatible.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Imports ActorExecutionTrace.java

Signed-off-by: Stefan Marr <git@stefan-marr.de>

clementbera committed:
asyncstackTrace structure working

clementbera committed:
asyncstackTrace structure working

clementbera committed:
update for potential integration

clementbera committed:
update for potential integration

clementbera committed:
som typo fix

clementbera committed:
som typo fix

clementbera committed:
default setting is now to true

clementbera committed:
default setting is now to true

clementbera committed:
fixed dynamic metric: (1) removed null hack (2) do not profile shadowStackEntry argument

clementbera committed:
fixed dynamic metric: (1) removed null hack (2) do not profile shadowStackEntry argument

smarr committed:
Fix initialization of transactional AtPut nodes

Didn’t indicate whether they are wrapped.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

clementbera committed:
Added cache logic. Disabled all analysis

clementbera committed:
Added cache logic. Disabled all analysis

clementbera committed:
rm print

clementbera committed:
rm print

smarr committed:
Generalize PrimitiveArgument to ArgumentExpr

This avoids the extra level of dynamicity needed in the instrumentation framework, which is not provided by Truffle.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
For uncaught exceptions, announce stack trace, and warn that itself might crash

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Get the language object more consistently via helper

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Handle null values in argument arrays for profiled operations

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove unnecessary cast in ArrayAllocationProfilingNode

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Added README, .gitignore, and checkstyle for extension tests

- clean test extension on clean

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Mark ExtensionPrims as abstract

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Make sure DummyParent has language when needed

- also add helper to access this of node in Truffle’s guard DSL

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
ClassSlotAccessNode: use guard in the node

Instead of accessing the read/write node for its guard, use the one cached in the node.

On construction we take this guard for easy access already. Using it directly from the node is problematic when wrappers are applied.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Updated expected dym results

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update superinstructions, all ? gone, not sure why

Didn’t investigate further, but this might indicate an issue.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove redundant check and fix compiler warning about raw type

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Handle methods that don’t have outer scope

These should be only superclass resolution methods.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix swapping of tracing buffers for Kompos

- use proper methods instead of public field to indicate the swap request
- coordinate with debugger. If a thread is blocked in a breakpoint (which triggers a trace data request) then we need to nicely swap the buffer ourselves. That’s safe of course.
- threads need to be registered to tracing backend also for Kompos

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add id for STM thread as debugging aid

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Output extra data for failing tests for basic protocol

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Need to configure debugger with our tags

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
ResolvePromiseNode has a HaltPrim, needs to be wrapped

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Comment out @SuppressWarnings("deprecation") because of javac bug

Signed-off-by: Stefan Marr <git@stefan-marr.de>

clementbera committed:
Rewrote the shadow stack logic to have caches in nodes, added shadow stack entries for lambdas

clementbera committed:
Rewrote the shadow stack logic to have caches in nodes, added shadow stack entries for lambdas

clementbera commented on src/tools/asyncstacktraces/ShadowStackEntry.java:
```diff
@@ -0,0 +1,62 @@
+
+  protected final ShadowStackEntry previousEntry;
+  protected final ExpressionNode   expression;
+  public static long               numberOfAllocations = 0;
```
Hum, that is an analysis only variable... Not sure it makes sense to look into this.

smarr replied:
Unclear why you initialize `numberOfAllocations` and `ALLOCATION_COUNT ` explicitly. `0` and `false` are the default values.

Also, for style, it would be nice to have a free line before `numberOfAllocations` to group the two separate things (instance vs class state) visually.

clementbera replied:
I will remove the initialization to zero, that will keep codacy happy. The other one I keep because I modify it frequently... I can make it a VM setting
 but I do not want to add a crazy number of settings.

smarr replied:
What's the problem with making it a proper `VmSetting`? Makes things much easier to use, no?
Also, you don't need to add it to the launcher. A simple `./som -Dsom.someSetting=true ...` will work fine.

smarr committed:
Remove unused ExpressionNode.assertIsStatement()

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Replaced removed Profiler with CPUSampler

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Adopt new Source builder

- this requires us to work a little more to get a proper path
  otherwise, we have an URI with a `file:` prefix.

- added factory method for source to SomLanguage

- added asserts in TypeGrammarParserTest as documentation

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Adopt find() methods on instruments

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Restructure VM startup to account for Truffle changes

The PolyglotEngine is so, we switch to Graal SDK.

- introduce Launcher, which is the new entry point for the execution
- START source is now in Launcher
- VM object is now created by createContext, we get access to the Context later on from inside the execution. So this should work, I hope.
- do shutdown() in disposeContext()
- Use SDK Option framework, currently only to support basic interpreter tests
  - for convenience allow testSelector to be empty string. Otherwise, the OptionsKey would need a custom implementation

- restructure shutdown/exit, simply return exit code
- don’t need support for avoidExitForTesting anymore

- adapted tests to new launcher
  - this requires some reflection to get to the real value for our tests
  - make sure Context is initialized for parallel BasicInterpreterTests
    - the context initialization is not thread safe.
      Not sure it should be.
      Use similar approach to Launcher.START to simply initialize the context but do nothing else.

- moved initialization of tools into context creation
  We need it earlier, because we are missing instrumentation on the standard library otherwise.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove TruffleSomTCK, because TruffleTCK is gone

It is replaced with the Polyglot TCK: https://github.com/oracle/graal/blob/master/truffle/docs/TCK.md

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[WIP] Add support for Polyglot TCK

This is work in progress, a LanguageProvider is a first step.

I think, the best way forward would be to allow parsing of top-level blocks.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

clementbera committed:
added Kernel.ns extra

clementbera committed:
added Kernel.ns extra

clementbera committed:
disabled 1 test to have bench results

clementbera committed:
disabled 1 test to have bench results

clementbera committed:
fixed bug in valueWithArgs

clementbera committed:
fixed bug in valueWithArgs

clementbera committed:
disable failing test

clementbera committed:
disable failing test

smarr committed:
Use Env instead of engine where possible

- this is more directly, and preferable over the access of the static context to get the engine

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update CoverallsTruffle for latest Truffle

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remember how we did debugging of javac

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[WIP] Don’t have any histogram or anything for the truffle profiler yet

Signed-off-by: Stefan Marr <git@stefan-marr.de>

clementbera committed:
re-enabled DNU tests and associated fixes

clementbera committed:
re-enabled DNU tests and associated fixes

smarr committed:
Update CI settings

- update Travis JDK versions and setup
- set JAVA_HOME for GitLab CI

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update codespeed.conf for ReBench 1.0

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use SIGKILL to terminate SOMns after test

Not sure why, but we seem to need a little more force.
Otherwise, the tests hang after completion.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
For debugging: Mark unused field in Parser

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Deprecations: Change slot.getKind to access via descriptor

This change needs to be revised later when we address issue #240.
We assume here that slots belong to one descriptor.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Removed unused som.actorTracingCfg handling

This was a leftover from an earlier incomplete change.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Ignore deprecations of GraphPrintVisitor for now

There is no replacement, let’s keep using it as long as possible.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

daumayr committed:
Fix teplay test flakiness

Exit code was changed to 1 when there were some unprocessed but expected messages at termination.

smarr committed:
Simplify changes

- need only one primitive
- initialize timer as part of the actor module init
- avoid use of method references (requires lambda infrastructure)
- use same style for the fields
- initialize structures only in replay mode
- change naming to be more explicit about semantics

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Apply todo, and solve a few compilation issues

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix BlockPrims

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix BlockPrims

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
@clementbera I rebased the branch on the latest `dev` branch, and fixed an issue with includes introduced during rebasing.

smarr committed:
Fixup Kompos stack rendering

- no radio input
- font size can be small

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fixup Kompos stack rendering

- no radio input
- font size can be small

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use shadow stack to satisfy Kompos stack trace requests

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use shadow stack to satisfy Kompos stack trace requests

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update bootstrap

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update bootstrap

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Mark async sends, or more generally actor switch points explicitly in stack

- actually, mark all elements without source location
- this means that we have an extra stack trace entry for actor transitions

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Mark async sends, or more generally actor switch points explicitly in stack

- actually, mark all elements without source location
- this means that we have an extra stack trace entry for actor transitions

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Track calling context also for promise callbacks

This is a first attempt to track the calling context for callbacks on promises.
It’s not quite right yet.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Track calling context also for promise callbacks

This is a first attempt to track the calling context for callbacks on promises.
It’s not quite right yet.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Ported ReplayDemo.som to ReplayDemo.ns

This example was originally implemented by Dominik Aumayr for his master thesis work.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Ported ReplayDemo.som to ReplayDemo.ns

This example was originally implemented by Dominik Aumayr for his master thesis work.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

daumayr committed:
first draft

daumayr committed:
first draft

smarr committed:
[HACK] Kompos: don’t fill logs with warnings

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[HACK] Kompos: don’t fill logs with warnings

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Lift restriction on shadow stack entry type

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Lift restriction on shadow stack entry type

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add asserts on shadow stack entries

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add asserts on shadow stack entries

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[WIP] All promise resolution/scheduling needs shadow stack entry

This needs more work, we probably need to put the shadow stack entry into the promise, which may be a problem for optimization.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[WIP] All promise resolution/scheduling needs shadow stack entry

This needs more work, we probably need to put the shadow stack entry into the promise, which may be a problem for optimization.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use named constants to access promise message args

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use named constants to access promise message args

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[WIP] Add extra node for Promise Resolution

Avoid surfacing breakpoint stuff and shadow stack, this needs to be handled in the ResolvePromiseNode, I think.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[WIP] Add extra node for Promise Resolution

Avoid surfacing breakpoint stuff and shadow stack, this needs to be handled in the ResolvePromiseNode, I think.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[WIP] TODOs

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[WIP] TODOs

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Refactor ShadowStackEntry hierarchy

- enforce more encapsulation, use factory methods instead of exposing classes
- add extra entry class for promise resolution
- rename async one to be more explicit what it is
- make them inner classes of ShadowStackEntry to have code more localized

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Refactor ShadowStackEntry hierarchy

- enforce more encapsulation, use factory methods instead of exposing classes
- add extra entry class for promise resolution
- rename async one to be more explicit what it is
- make them inner classes of ShadowStackEntry to have code more localized

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Make stack frames clickable in Kompos

Move getSectionIdFromFrame out of debugger, seems more general. locate it next to getSectionId in messages.ts.

Make all elements that have a source location clickable.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Make stack frames clickable in Kompos

Move getSectionIdFromFrame out of debugger, seems more general. locate it next to getSectionId in messages.ts.

Make all elements that have a source location clickable.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[HACK?] Take context when resolving promise from the stack

- need to pass the frames accordingly

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[HACK?] Take context when resolving promise from the stack

- need to pass the frames accordingly

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Implement StackIterator to encapsulate all iteration logic

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Implement StackIterator to encapsulate all iteration logic

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Change default port for HTTP to 8880

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Change default port for HTTP to 8880

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[REM] Added todo for assertions that we should have

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[REM] Added todo for assertions that we should have

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[HACK] Demo: disable unnecessary visualization

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[HACK] Demo: disable unnecessary visualization

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Unify stack trace handling: use StackIterator for everything

- this removes the asyncError: and PrintAsyncStackTracePrim
- generalize some MaterializedFrame to Frame (simplest change, but may mean we inadvertently see VirtualFrame objects

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Unify stack trace handling: use StackIterator for everything

- this removes the asyncError: and PrintAsyncStackTracePrim
- generalize some MaterializedFrame to Frame (simplest change, but may mean we inadvertently see VirtualFrame objects

Signed-off-by: Stefan Marr <git@stefan-marr.de>

clementbera committed:
clean up unused stuff and moved to fixed version

clementbera committed:
clean up unused stuff and moved to fixed version

clementbera committed:
final clean up: only most efficient and simpler version kept

clementbera committed:
final clean up: only most efficient and simpler version kept

clementbera committed:
Added method based cache which backpoints to the CachedDispatchNode / the LexicallyBoundDispatchNode. Wrote a printAsyncStackTrace method temporarily to check it works. Fib just worked without Graal enabled. Next step are to change cast to support HotspotOptimizedCallTarget & wrote a stackIterator dealing with method cache from the printAsyncStackTrace method. I need to fix asynchronous calls too

clementbera committed:
Added method based cache which backpoints to the CachedDispatchNode / the LexicallyBoundDispatchNode. Wrote a printAsyncStackTrace method temporarily to check it works. Fib just worked without Graal enabled. Next step are to change cast to support HotspotOptimizedCallTarget & wrote a stackIterator dealing with method cache from the printAsyncStackTrace method. I need to fix asynchronous calls too

clementbera committed:
fixed the problem with HotspotOptimizedTarget. Refactored a bit to avoid code duplication

clementbera committed:
fixed the problem with HotspotOptimizedTarget. Refactored a bit to avoid code duplication

clementbera committed:
Removed requiresShadowStack from the logic: inline cache to primitive requires a shadow stack entry since unlike Smalltalk primitive can be reentrant, typically ResolvePrimitiveGen can trigger a send. Added the extra arguments for ShadowStackEntry in many different new places where send arguments are hand written. Add new assertions.

clementbera committed:
Removed requiresShadowStack from the logic: inline cache to primitive requires a shadow stack entry since unlike Smalltalk primitive can be reentrant, typically ResolvePrimitiveGen can trigger a send. Added the extra arguments for ShadowStackEntry in many different new places where send arguments are hand written. Add new assertions.

clementbera committed:
make the first SSEntry always an SSEntry instead of null. It made me realize at some points the SSEntry broke (due to null entries). It is now explicit with instantiateTopEntry.

clementbera committed:
make the first SSEntry always an SSEntry instead of null. It made me realize at some points the SSEntry broke (due to null entries). It is now explicit with instantiateTopEntry.

clementbera committed:
added settings in som for method cache and inline cache. Make it work again with async stack trace enabled. I am stuck with Ensure/OnDo specialization error. It comes form having promises instead of Block. I need to investigate

clementbera committed:
added settings in som for method cache and inline cache. Make it work again with async stack trace enabled. I am stuck with Ensure/OnDo specialization error. It comes form having promises instead of Block. I need to investigate

clementbera committed:
fix bug in argument handling in som. Try to make SOMns work with async-stack-traces disabled, but I have the same crash with having a promise in ensure having a receiver being a promise and not a block. I need to check diffs to understand where it comes from, debugging does not show related errors.

clementbera committed:
fix bug in argument handling in som. Try to make SOMns work with async-stack-traces disabled, but I have the same crash with having a promise in ensure having a receiver being a promise and not a block. I need to check diffs to understand where it comes from, debugging does not show related errors.

clementbera committed:
removed debugging code. Fix additional case where there was a depedency on ShadowStackEntry type, breaking the code where Async stack trace is disabled. Still trying to fix the async calls which makes the VM crash on exception specialization with promise instead of block parameters

clementbera committed:
fixed more ShadowStackEntry references in non async stack trace code. REmoved debugging code.

clementbera commented:
Moved to async stack trace (from dev) PR.
